substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,548 @@
+        for (InlineInvokePlugin inlineInvokePlugin : originalGraphBuilderConfig.getPlugins().getInlineInvokePlugins()) {
+            if (!(inlineInvokePlugin instanceof NativeImageInlineDuringParsingPlugin)) {
+                result.getPlugins().appendInlineInvokePlugin(inlineInvokePlugin);
+            }
+        }
+
+        return result;
+    }
+
+    @SuppressWarnings("try")
+    InvocationResult analyzeMethod(CallSite callSite, AnalysisMethod method, ValueNode[] args, Object existingSingleAllowedElement) {
+        if (!method.hasBytecodes()) {
+            /* Native method. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+
+        } else if (method.isSynchronized()) {
+            /*
+             * Synchronization operations will always bring us above the node limit, so no point in
+             * starting an analysis.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+
+        } else if (providers.getGraphBuilderPlugins().getInvocationPlugins().lookupInvocation(method) != null) {
+            /* Method has an invocation plugin that we must not miss. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+
+        MethodState methodState = new MethodState(new InvocationResultInline(callSite, method), existingSingleAllowedElement);
+
+        GraphBuilderConfiguration graphBuilderConfig = prototypeGraphBuilderConfig.copy();
+        graphBuilderConfig.getPlugins().appendInlineInvokePlugin(methodState);
+        graphBuilderConfig.getPlugins().appendParameterPlugin(new TrivialMethodDetectorParameterPlugin(args));
+
+        StructuredGraph graph = new StructuredGraph.Builder(options, debug).method(method).build();
+
+        try (DebugContext.Scope ignored = debug.scope("InlineDuringParsingAnalysis", graph, method, this)) {
+
+            TrivialMethodDetectorGraphBuilderPhase builderPhase = new TrivialMethodDetectorGraphBuilderPhase(bb, providers, graphBuilderConfig, OptimisticOptimizations.NONE, null,
+                    providers.getWordTypes());
+
+            try (NodeEventScope ignored1 = graph.trackNodeEvents(methodState)) {
+                builderPhase.apply(graph);
+            }
+
+            debug.dump(DebugContext.BASIC_LEVEL, graph, "InlineDuringParsingAnalysis successful");
+            return methodState.analyzeGraph();
+        } catch (Throwable ex) {
+            debug.dump(DebugContext.BASIC_LEVEL, graph, "InlineDuringParsingAnalysis failed with " + ex.toString());
+            /*
+             * Whatever happens during the analysis is non-fatal because we can just not inline that
+             * invocation.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+    }
+
+    /**
+     * We collect information during graph construction using and use result of
+     * {@link #analyzeGraph} in {@link #analyzeMethod}.
+     */
+    class MethodState extends NodeEventListener implements InlineInvokePlugin {
```

Comment:
I find it very confusing that MethodState is three things at the same time. Maybe give it a better name or simply split it up in two things that get state as input.

Reply:
Renamed MethodState to MethodNodeTrackingAndInline.